Source/Scene/Cesium3DTilesetTraversal.js

```diff
@@ -1,856 +1,643 @@
-                var child = isNativeArray ? children[i] : children.get(i);
-
-                if (!defined(options.shouldVisit) || options.shouldVisit(child)) {
-                    stack.push(child);
+            if (ancestorStack.length > 0) {
+                var waitingTile = ancestorStack.peek();
+                if (waitingTile._stackLength === stack.length) {
+                    ancestorStack.pop();
+                    if (waitingTile === lastAncestor) {
+                        waitingTile._finalResolution = true;
+                    }
+                    selectTile(tileset, waitingTile, frameState);
+                    continue;
                 }
             }
 
-            if (length === 0 && defined(options.leafHandler)) {
-                options.leafHandler(tile);
+            var tile = stack.pop();
+            if (!defined(tile)) {
+                // stack is empty but ancestorStack isn't
+                continue;
             }
-            options.visitEnd(tile);
-        }
-
-        stack.trim(maxLength);
-    }
 
-    function breadthFirstSearch(root, options) {
-        var queue1 = options.queue1;
-        var queue2 = options.queue2;
-
-        if (defined(root) && (!defined(options.shouldVisit) || options.shouldVisit(root))) {
-            queue1.push(root);
-        }
+            var add = tile.refine === Cesium3DTileRefine.ADD;
+            var shouldSelect = tile._shouldSelect;
+            var children = tile.children;
+            var childrenLength = children.length;
+            var traverse = (childrenLength > 0) && (tile._screenSpaceError > maximumScreenSpaceError);
 
-        var maxLength = 0;
-        while (queue1.length > 0) {
-            var length = queue1.length;
-            maxLength = Math.max(maxLength, length);
-
-            for (var i = 0; i < length; ++i) {
-                var tile = queue1.get(i);
-                options.visitStart(tile);
-                var children = options.getChildren(tile);
-                var isNativeArray = !defined(children.get);
-                var childrenLength = children.length;
-                for (var j = 0; j < childrenLength; ++j) {
-                    var child = isNativeArray ? children[j] : children.get(j);
-
-                    if (!defined(options.shouldVisit) || options.shouldVisit(child)) {
-                        queue2.push(child);
+            if (shouldSelect) {
+                if (add) {
+                    selectTile(tileset, tile, frameState);
+                    tile._finalResolution = true;
+                } else {
+                    tile._selectionDepth = ancestorStack.length;
+                    if (tile._selectionDepth > 0) {
+                        tileset._hasMixedContent = true;
+                    }
+                    lastAncestor = tile;
+                    if (!traverse) {
+                        selectTile(tileset, tile, frameState);
+                        tile._finalResolution = true;
```

Comment:
So `_finalResolution` is a tile marked for rendering at the end of this round of traversal, correct? I think we should call this `_selectedForRender` or something a little more specific (as it is used externally to this class for debug code).

Reply:
`_selectedForRender` is a bit too broad, `_finalResolution` says a tile is rendered AND none of its descendants are rendered. I wonder there's a better name specifically for that condition...

But that's where the `!finalResolution -> yellow` comes in, it indicates a tile and its descendant are rendering simultaneously and the tile is not entirely "resolved".

I did clean up some of the `_finalResolution` code though so that `_finalResolution = false` is set in one place instead of `_finalResolution = true` being set everywhere.